autoload/ale/completion.vim

```diff
@@ -183,52 +185,71 @@ function! ale#completion#RestoreCompletionOptions() abort
     endif
 endfunction
 
+function! ale#completion#FindCompletionStart() abort
+    let [l:line, l:column] = getcurpos()[1:2]
+    let l:regex = s:GetFiletypeValue(s:omni_start_map, &filetype)
+    let l:up_to_column = getline(l:line)[: l:column - 2]
+    let l:match = matchstr(l:up_to_column, l:regex)
+
+    let b:ale_completion_start = l:column - len(l:match) - 1
+
+    return b:ale_completion_start
+endfunction
+
+function! ale#completion#PollCompletionResults() abort
+    if ale#completion#Queue()
+        return []
+    else
+        return get(b:, 'ale_completion_result', [])
+    endif
+endfunction
+
 function! ale#completion#OmniFunc(findstart, base) abort
     if a:findstart
-        let l:line = b:ale_completion_info.line
-        let l:column = b:ale_completion_info.column
-        let l:regex = s:GetFiletypeValue(s:omni_start_map, &filetype)
-        let l:up_to_column = getline(l:line)[: l:column - 2]
-        let l:match = matchstr(l:up_to_column, l:regex)
-
-        return l:column - len(l:match) - 1
+        return ale#completion#FindCompletionStart()
     else
-        " Parse a new response if there is one.
-        if exists('b:ale_completion_response')
-        \&& exists('b:ale_completion_parser')
-            let l:response = b:ale_completion_response
-            let l:parser = b:ale_completion_parser
+        return ale#completion#PollCompletionResults()
+    endif
+endfunction
 
-            unlet b:ale_completion_response
-            unlet b:ale_completion_parser
+function! ale#completion#QueryDone() abort
+    let s:last_done_pos = getcurpos()[1:2]
 
-            let b:ale_completion_result = function(l:parser)(l:response)
-        endif
+    let b:ale_completion_result = ALEProcessCompletionResults(get(b:, 'ale_completion_result', []))
 
-        call s:ReplaceCompletionOptions()
+    call ALEAfterCompletionResults()
+endfunction
 
-        return get(b:, 'ale_completion_result', [])
-    endif
+function! ALEProcessCompletionResults(completion_results) abort
+    " Users may override this function to act on the results however they
+    " see fit. Needs to be not namespaced.
+    return a:completion_results
 endfunction
 
-function! ale#completion#Show(response, completion_parser) abort
-    if ale#util#Mode() isnot# 'i'
+function! ALEAfterCompletionResults() abort
```

Comment:
Doesn't deoplete accept autoload functions? Why not?

Reply:
Does deoplete need to use this function at all? This function opens up the completion menu itself, so I'd imagine it's only useful when you aren't using deoplete.

Reply:
When a result is parsed into menu candidates, after that, the caller needs a way to know that the results are ready to be used on their end. For deoplete I've overrode this function with:

```
function! ALEAfterCompletionResults() abort
    call deoplete#send_event('CompleteDone', 'ale')
endfunction
```

For other completion plugins, people can override this "call be when it's done callback" with something else, hence the name _**After**_CompletionResults.

Reply:
Replacing an existing function like that is bad design, and it doesn't make sense. Instead of doing that, let's add the deoplete source to ALE, and we don't need to call the function for opening the omnicomplete menu when ALE completion integration with LSP is being used as a completion source.